.github/workflows/python.yml

```diff
@@ -95,30 +95,39 @@ jobs:
       - name: Test
         working-directory: py/tests
         run: python3 -m pytest
+      - name: Display structure of dist directory
+        run: ls -R dist
       - name: Upload wheels
-        if: "startsWith(github.ref, 'refs/tags/')"
+        # if: "startsWith(github.ref, 'refs/tags/')"
         uses: actions/upload-artifact@v4
         with:
-          name: wheels
+          # Ensure we have a different name for every artifact, by using the
+          # matrix components in the artifact name
+          name: binary-${{ matrix.type }}-${{ matrix.os }}-${{ matrix.python-version }}
           path: dist
 
   release:
     name: Release
     runs-on: ubuntu-latest
-    if: "startsWith(github.ref, 'refs/tags/')"
     needs: [build]
     environment: VALIDATOR_RELEASE
     steps:
       - uses: actions/download-artifact@v4
         with:
-          name: wheels
+          pattern: binary-*
```

Comment:
I think we should be more specific here to prevent downloading artifacts from the non-tag workflow run. I'm unsure what would happen to the duplicates if we do it like this.

Reply:
My understanding of Github Artifacts is that artifacts are shared _within_ a workflow, but not _across_ workflows - so that while the branch workflow and tag workflow might have artifacts with the same name, they would not be able to see each other.

From [the docs](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/storing-and-sharing-data-from-a-workflow):

> Artifacts allow you to persist data after a job has completed, and share that data with another job in the same workflow. 

See also the [`download-artifact`](https://github.com/actions/download-artifact/blob/main/README.md#download-artifacts-from-other-workflow-runs-or-repositories) docs, which discuss adding a tokens and run id to the `download-artifact` action to enable downloading from other runs (which we don't want to do).